neuralpp/test/quick_tests/symbolic/expression_test.py

```diff
@@ -186,17 +185,49 @@ def test_function_application(expression_factory):
 
-    fa3 = fa.set(0, SymPyExpression.new_constant(operator.mul, int_to_int_to_int))
-    assert fa3 == SymPyExpression.new_function_application(SymPyExpression.new_constant(operator.mul,
-                                                                                        int_to_int_to_int),
+    fa3 = fa.set(0, expression_factory.new_constant(operator.mul, int_to_int_to_int))
+    assert fa3 == expression_factory.new_function_application(expression_factory.new_constant(operator.mul,
+                                                                                              int_to_int_to_int),
                                                            [constant_one, constant_two])
-    assert fa == SymPyExpression.new_function_application(add_func, [constant_one, constant_two])
-    fa6 = fa.set(1, SymPyExpression.new_variable("a", int))
-    assert fa6 == SymPyExpression.new_function_application(add_func, [SymPyExpression.new_variable("a", int),
-                                                                      constant_two])
+    assert fa == expression_factory.new_function_application(add_func, [constant_one, constant_two])
+    fa6 = fa.set(1, expression_factory.new_variable("a", int))
+    assert fa6 == expression_factory.new_function_application(add_func, [expression_factory.new_variable("a", int),
+                                                                         constant_two])
     with pytest.raises(IndexError):
         fa2.set(3, constant_one)
 
+    # operator overloading
+    x = expression_factory.new_variable("x", int)
+    x_plus_one = x + 1
+    assert x_plus_one.function.type == int_to_int_to_int
+    x_minus_one = x - 1
+    assert x_minus_one.function.type == int_to_int_to_int
+    if expression_factory != Z3Expression:
+        # these tests won't work in Z3 because it does not support arithmetic of different types
+        one_third_plus_x = fractions.Fraction(1, 3) + x
+        assert one_third_plus_x.function.type == Callable[[fractions.Fraction, int], fractions.Fraction]
+        pi_times_x = math.pi * x
+        assert pi_times_x.function.type == Callable[[float, int], float]
+
+    b1 = expression_factory.new_variable("b1", bool)
+    b2 = expression_factory.new_variable("b2", bool)
+    b3 = expression_factory.new_variable("b3", bool)
+    expr = b1 & b2 | ~b3
+    assert expr.function.type == Callable[[bool, bool], bool]  # or
+    if expression_factory != SymPyExpression:
+        assert expr.arguments[0].function.type == Callable[[bool, bool], bool]  # and
+        assert expr.arguments[1].function.type == Callable[[bool], bool]  # not
+        expr1 = b1 & True
+        assert expr1.function.type == Callable[[bool, bool], bool]
+    else:  # sympy changes the argument order
+        assert expr.arguments[0].function.type == Callable[[bool], bool]  # not
+        assert expr.arguments[1].function.type == Callable[[bool, bool], bool]  # and
```

Comment:
Similar as above, sympy internally change the argument order so it would also be a bit surprising to a user.

Reply:
Yes, should be fine for now too.